### PR TITLE
kicad.libraries.packages3d:  reduce size to allow caching

### PR DESCRIPTION
**Pull request by evils**
## Description of changes

add the `stepreduce` package and use it to reduce the size of the step files in the KiCad 3D models
and apply zip compression for .stpZ output

these bring the closure size bellow 2G allowing hydra to cache the output
(downloading the packages from gitlab directly sometimes fails, and it takes a while which leads some people to think something went wrong)

stepreduce brings down closure size from 5.4G to 3.3G
zip brings it down from 3.3G to 1.6G (or 5.4G to 1.9G without reduce)

xz compressed NAR sizes are:
plain: 373M
gzip: 712M
zip: 712M
stepreduce: 262M
stepreduce & gzip: 465M
(gzip only works here coincidentally because our KiCad was built with a library that has that support
[the quasi standard of stpZ is more universally supported](https://www.mbx-if.org/documents/rec_prac_file_compression_v13.pdf))
stepreduce & zip: 468M

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/science/electronics/kicad/default.nix [hunk 1]
```diff
@@ -229,18 +229,9 @@ stdenv.mkDerivation rec {
...
     '';
     license = lib.licenses.gpl3Plus;
-    maintainers = with lib.maintainers; [ evils kiwi ];
-    # kicad is cross platform
+    maintainers = with lib.maintainers; [ evils ];
     platforms = lib.platforms.all;
     broken = stdenv.isDarwin;
-
-    hydraPlatforms = if (with3d) then [ ] else platforms;
-    # We can't download the 3d models on Hydra,
```
Comment (evils) [inline pkgs/applications/science/electronics/kicad/default.nix:238]: this comment makes it sound like the downloading itself is the issue
TBD if this actually works on hydra now...

Comment (evils) [reply pkgs/applications/science/electronics/kicad/default.nix:238]: derivation output size (total of all outputs) is limited to 3G, this change should work

**Comment by bjornfor** [comment]
Looking forward to having this cached, thanks!

> [kicad: remove kiwi maintainer](https://github.com/NixOS/nixpkgs/pull/248517/commits/da346d9a004243d9969d76bb9564eedab552ce3e)

Can you comment why in the commit message?

Is stepreduce really needed / fully safe? If I read it right, it only takes the final output from 1.9 to 1.6 GiB. IMHO, that's not so big difference compared to the size reduction due to compression. The [code](https://gitlab.com/sethhillbrand/stepreduce/-/blob/master/stepreduce.cpp?ref_type=heads) seems simple enough, but (1) it has no release and (2) it's only packaged in archlinux (AUR) so far: https://repology.org/project/stepreduce/versions. (This is not me rejecting but just telling my current opinion and asking for feedback.)

**Comment by evils** [comment]
> Can you comment why in the commit message?

done, thanks for pointing that out

> Is stepreduce really needed / fully safe?

i think the advantage it brings is worth it
(note that it makes the compressed NAR around 35% smaller vs just zip)
the author has done some work verifying its output
https://archive.fosdem.org/2020/schedule/event/step_reduce/
(and also points out that this makes KiCad's board outputs smaller too)

i believe the reason for the lack of adoption is mostly that people (including me) forgot about it or were too busy to apply it (i think the KiCad library team consisted of 0 people at some point)

i've asked the author if he could tag a release and got a positive response
but between managing the KiCad project and next month's KiCon,
i suspect he's got a few more urgent things to attend to :)

**Comment by bjornfor** [comment]
> xz compressed NAR sizes are:
plain: 373M
gzip: 712M
zip: 712M
stepreduce: 262M
stepreduce & gzip: 465M
stepreduce & zip: 468M

Whys is plain smaller than gzip etc.? Typo?

**Comment by evils** [comment]
> Why is plain smaller than gzip etc.?

because that listing is with xz compression as used by Nix, which apparently doesn't do well on already compressed zip files
(i assume zip removes some redundancy that xz could have used more effectively)

File: pkgs/applications/misc/stepreduce/default.nix [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  version = "unstable-2020-04-30";
+
+  src = fetchFromGitLab {
+    owner = "sethhillbrand";
+    repo = "stepreduce";
+    rev = "e89091c33b67e2a18584e1fe3560bfd48ae98773";
+    hash = "sha256-bCseBQ6J3sWFt0kzaRkV11lwzOGvNPebvQ6w4OJaMBs=";
+  };
+
+  installPhase = "install -Dm755 stepreduce $out/bin/stepreduce";
```
Comment (bjornfor) [inline pkgs/applications/misc/stepreduce/default.nix:17]: Please include pre/post hooks:

```
  installPhase = ''
    runHook preInstall

    install -Dm755 stepreduce $out/bin/stepreduce

    runHook postInstall
  '';
```

**Comment by bjornfor** [comment]
> kicad: remove kiwi maintainer
>
> they're no longer active

Nit: Commit message body (as opposed to the title) should be capitalized and punctuated.

**Comment by evils** [comment]
not according to my style :)

**Comment by bjornfor** [comment]
Whats your style guide? I don't know if nixpkgs has its own, but I think https://cbea.ms/git-commit/#seven-rules is generally considered good.

**Comment by evils** [comment]
i don't have a style guide
but i'm never going to spontaneously capitalize anything and want to keep my commits consistent

**Comment by evils** [comment]
sorry, i should have been more verbose with my initial response

i think debating style at this level is going to be productive
(meaning either we enforce it programmatically or make do with good enough)

File: pkgs/applications/misc/stepreduce/default.nix (hunk 2 above)
Comment (evils) [reply pkgs/applications/misc/stepreduce/default.nix]: woeps, i let the the style discussion distracted me from this
sorry for the late response, i've included the suggestion, thanks
(it's such an easy thing to miss, is there a better way to do this?)

**Review by bjornfor — APPROVED** [review]

**Comment by evils** [comment]
thanks!

**Pull request merged**